core/src/main/java/org/apache/iceberg/rest/auth/AuthManager.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.rest.auth;
+
+import java.util.Map;
+import java.util.function.Supplier;
+import org.apache.iceberg.rest.RESTClient;
+import org.apache.iceberg.util.Pair;
+
+public interface AuthManager extends AutoCloseable {
+
+  Map<String, String> mergeAuthHeadersForGetConfig(
```

Comment:
This can be improved. Here we are constructing an initial auth session so a better signature would be:

```java
  AuthSession initialAuth(RESTClient initialAuthClient, Map<String, String> initialHeaders);
```

It's very easy to adapt the only call site of that method.

Reply:
I'd say it's actually the opposite: we need more methods.

In my exploration I came up with the following interface which works well:

```java
  AuthSession newInitialSession(RESTClient initialAuthClient, Map<String, String> initialHeaders);

  AuthSession newCatalogSession(
      RESTClient authClient, Map<String, String> properties, Map<String, String> catalogHeaders);

  AuthSession newContextualSession(
      SessionCatalog.SessionContext context, AuthSession parentSession);

  AuthSession newTableSession(Map<String, String> properties, AuthSession parentSession);
```

We need a different factory method for each "level" (or "scope") where an auth session could be introduced (and cached).

Reply:
However I agree that the first method `newInitialSession` is really only meant to produce a short-lived, temporary auth session in order to call the config endpoint. It's discarded immediately after. But I think we need it and we can't use `newCatalogSession` in this case.